### PR TITLE
Core: Add `globals` URL param and remove from sessionStorage

### PR DESCRIPTION
**Pull request by ghengeveld**
Issue: #11604

## What I did

Added `globals` URL query param that tracks any globals that have a value that deviates from their initial/default value.
Removed session storage for globals.

## How to test

- Is this testable with Jest or Chromatic screenshots? no
- Does this need a new example in the kitchen sink apps? no
- Does this need an update to the documentation? yes, todo

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: lib/client-api/src/types.ts [hunk 1]
```diff
@@ -39,6 +39,7 @@ export interface StoreSelectionSpecifier {
   viewMode: ViewMode;
   singleStory?: boolean;
   args?: Args;
+  globalArgs?: Args;
```
Comment (shilman) [inline lib/client-api/src/types.ts:42]: We renamed `globalArgs` to `globals` everywhere; I suggest we don't add it back now. cc @tmeasday

**Review by shilman — CHANGES_REQUESTED** [review]
Looking great w/ one suggestion! 💯

File: lib/client-api/src/story_store.test.ts [hunk 2]
```diff
@@ -509,16 +498,17 @@ describe('preview.story_store', () => {
     });
 
     it('sensibly re-initializes with memory based on session storage', () => {
-      (store2.session.get as any).mockReturnValueOnce({
-        globals: {
+      const store = new StoryStore({ channel });
+      store.setSelectionSpecifier({
+        storySpecifier: '*',
+        viewMode: 'story',
+        globalArgs: {
```
Comment (tmeasday) [inline lib/client-api/src/story_store.test.ts:505]: `globalArgs`? I think that's a different thing.

File: lib/api/src/modules/url.ts [hunk 3]
```diff
@@ -178,6 +183,14 @@ export const init: ModuleFn = ({ store, navigate, state, provider, fullAPI, ...r
       }
     });
 
+    fullAPI.on(GLOBALS_UPDATED, ({ globals, defaultGlobals, initialGlobals }) => {
```
Comment (tmeasday) [inline lib/api/src/modules/url.ts:186]: I'm not sure why we store and emit the `defaultGlobals` and `initialGlobals` separately. They sort of seem like the same thing, coming from different places, and they are treated the same way one the manager side too, right?

**Review by tmeasday — CHANGES_REQUESTED** [review]
Couple questions

File: lib/client-api/src/types.ts (hunk 1 above)
Comment (ghengeveld) [reply lib/client-api/src/types.ts]: Wasn't aware of that. Renamed.

**Review by shilman — APPROVED** [review]
💯

File: lib/api/src/modules/url.ts (hunk 3 above)
Comment (ghengeveld) [reply lib/api/src/modules/url.ts]: Sure. That's fair. I integrated `defaultGlobals` into `initialGlobals` now, so we only have to pass that.

**Comment by nickofthyme** [comment]
@ghengeveld @shilman is there a way to hide or prevent global params from syncing to the url? I initially thought this was a good idea but after building my own addon, this locks-in the global variable for all stories.

Say I have a story that has a background color as a global parameter. Say the parameter default is only set in the `preview.js` file to `#FFFFFF` for `story1`. This will read the color from the default `parameters` from `preview.js` and update the url. Now if I navigate to `story2` that has a story-level parameter override with a default of `#000000`, the color is still set to `#FFFFFF` because of the query param.

I DO think the hierarchy is correct and the url param value should override any other defined value. But syncing the global values to the URL creates issues. Looking at the old knobs addon, the knob values where never written to the url but you could still copy the current knob state via the `copy` button, which would then provide a stateful url that includes the necessary query params.

I think the capture state method used with the knobs would be a better approach in the case of the globals. Is there a reason this was done this way and would you consider changing this?

**Comment by shilman** [comment]
cc @MichaelArestad ☝️

**Comment by ghengeveld** [comment]
> I think the capture state method used with the knobs would be a better approach in the case of the globals. Is there a reason this was done this way and would you consider changing this?

The rationale is that we wanted to make it super easy and obvious to share stories and their state with others. Syncing the URL makes it easy to just copy the URL from the address bar, regardless of how they got there. A separate `Copy` button would require feature discovery. We did consider a "share" button in the toolbar (not in the addon panel like Knobs has, because args and globals are a core feature), but decided against it for now. If URL syncing continues to be problematic we might reconsider. I'm pretty sure at that point we'll design a more elaborate "share" feature that gives you more than just a URL.

I think a short-term improvement would be to enable addon authors to opt-out their globals from URL syncing. Some globals aren't worth syncing with the URL, such as the measure addon state.

**Comment by nickofthyme** [comment]
> I think a short-term improvement would be to enable addon authors to opt-out their globals from URL syncing. Some globals aren't worth syncing with the URL, such as the measure addon state.

Yes this would be greatly appreciated

**Comment by ghengeveld** [comment]
https://github.com/storybookjs/storybook/issues/15756

**Pull request merged**